atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java

```diff
@@ -324,24 +336,92 @@ private synchronized void refreshPool() {
 
     @VisibleForTesting
     void setServersInPoolTo(ImmutableSet<CassandraServer> desiredServers) {
-        ImmutableSet<CassandraServer> cachedServers = getCachedServers();
-        SetView<CassandraServer> serversToAdd = Sets.difference(desiredServers, cachedServers);
-        SetView<CassandraServer> absentServers = Sets.difference(cachedServers, desiredServers);
+        Set<CassandraServer> currentServers = getCachedServers();
+        SetView<CassandraServer> serversToAdd = Sets.difference(desiredServers, currentServers);
+        SetView<CassandraServer> absentServers = Sets.difference(currentServers, desiredServers);
 
-        serversToAdd.forEach(server -> cassandra.returnOrCreatePool(server, absentHostTracker.returnPool(server)));
         absentServers.forEach(cassandraServer -> {
             CassandraClientPoolingContainer container = cassandra.removePool(cassandraServer);
             absentHostTracker.trackAbsentCassandraServer(cassandraServer, container);
         });
 
         Set<CassandraServer> serversToShutdown = absentHostTracker.incrementAbsenceAndRemove();
 
-        if (!(serversToAdd.isEmpty() && absentServers.isEmpty())) { // if we made any changes
-            sanityCheckRingConsistency();
+        Set<CassandraServer> validatedServersToAdd =
+                validateNewHostsTopologiesAndMaybeAddToPool(getCurrentPools(), serversToAdd);
```

Comment:
Why are we re-fetching the current pools? We could just get this once, then derive `currentServers` from that. I think this is mainly nice as it should protect us a little more from any sneaky race conditions

Reply:
Yes I agree, and there is a commit with this change, but it revealed a nasty bug: we must call `getCurrentPools()` again as we've modified the underlying pool (i.e. we removed the servers which are no longer used/present).

Reply:
Alternatively, we could just filter these hosts from our "snapshot"

Reply:
OK, I've just re-read the code, and the refresh codepath should be only ever run on a single thread at any one time. That means we should be resilient to any race conditions in this case.